Initial Crossgen2 changes for composite R2R support

trylek opened the pull request:
This change adds the new command-line option "--composite",
a new type of input files "--unrooted-input-file" for files that should
be made part of the composite image but only rooted as hit by the
dependency analysis, not automatically as the normal input files
(to reduce the size of framework built along with the app) and new
Crossgen2 logic for producing the additional R2R metadata needed
by composite images including rewriting of input MSIL by injecting
a "component R2R header" on them that forwards the native code
to the composite image. With a set of runtime changes out for
a separate PR I'm able to make first steps in running composite R2R
in the CoreCLR runtime.

Thanks

Tomas

trylek committed:
Initial Crossgen2 fixes for composite R2R support

With initial local changes to enable composite R2R I have finally
been able to compile a CoreCLR unit test in this mode. It doesn't
yet work at runtime, these changes just fix initial compiler issues
I hit during the build:

1) Delegate constructor signature was missing some comparison logic
that ended up asserting in the node comparer in the composite
build.

2) I have removed two seeminly superfluous parameters to
CreateMethodEntrypointNodeHelper.

3) I have removed an assertion related to multi-file RVA fields
that I believe to be no longer relevant after the refactorings
to copy the RVA fields around properly.

Thanks

Tomas

trylek committed:
Address Michal's PR feedback

I have verified locally that the explicit passing of
isInstantiatingStub to DelegateCtorSignature is indeed superflous,
it was a leftover from my previous attempts to fix the aliasing
problem that later turned out to be caused by the incomplete
comparison. I have rolled back this part of my change.

Thanks

Tomas

trylek committed:
WIP: Initial Crossgen2 changes for composite R2R support

This is currently a big WIP. This change lets me compile a test
together with the framework into a single composite R2R image.
With this change the image does get produced but is still incorrect
as the two-level header structure is not yet included so that
the available types and method entrypoints cannot be properly
resolved.

As next step I plan to add this support together with matching
R2RDump logic to let me statically prove that the produced image
is consistent before I start working on the matching runtime support.
Some general observations:

1) Currently this change includes my DelegateCtorSignature
comparison fix I have sent out for a separate review. Based on
complexity of the composite R2R format changes Ill either rebase
this PR once I merge the DelegateCtorSignature fix in or Ill
send out finalized parts of this change for separate PRs
and subsequently Ill close this change.

2) With all framework assemblies rooted a simple CoreCLR test
builds for about 10 minutes and the resulting executable has
about 142 MB. For now I have modified Program.cs so that, in
composite mode, reference assemblies (-r) are also added as
input files but arent automatically rooted for R2R compilation
i.o.w. we compile only their subset that is transitively reachable
from the input assemblies. With this policy the compiled test
executable has about 1 MB. This will definitely require more thought.

3) I had to temporarily delete a check in IBCProfileData that was
crashing in the composite build mode. Ill welcome any feedback
as to what is the right fix here.

4) I keep constantly using my local instrumentation to downgrade
SuperIlc to .NET Core 3.0 framework when building from the VS
as even the newest VS still seems unable to build .NET Core 5-targeting
apps. Ill welcome any thoughts as to whether this merits merging in.

5) For now I plan to modify the R2R PE emitter to stop copying over
the managed headers and parts of the input file in the composite mode
and emit the header from scratch. One interesting question here is
what to do about resource sections  Im not sure if its legal for
a PE executable to contain multiple .rsrc sections so wed either
need to copy around just the first one or somehow merge the resources.
Alternatively we may be able to ignore the resources with the rationale
that, at least in the standalone MSIL case, they remain in existence
in the original MSIL files, and just perhaps support specifying
an explicit resource file to attach to the executable (e.g. to
set up its file manager icon).

Thanks

Tomas

davidwrighton commented on src/coreclr/src/tools/crossgen2/crossgen2/Program.cs:
```diff
@@ -242,6 +244,18 @@ private int Run()
+                                // In composite mode, add reference assemblies as inputs but don't root all their methods
+                                // This will need more granularity (probably two types of references) when implementing
+                                // shared framework / multi-layer composite R2R files.
+                                inputModules.Add(module);
```
It seems very strange that we'd want to include in the input set any modules that were only specified as references. I don't see why you are doing this. Instead I would expect to see the logic that changed what were the reference vs input modules to be placed into SuperILC

trylek replied:
As I described elsewhere, this was more like a first shot at something like a potential rooting policy. Rooting the entire framework is simply impractical for testing as the build takes ages and the output file is huge; the limited rooting makes the build much easier to manage, at least for the purpose of local testing. If you think there's no value in having this knob, I can easily keep this as local instrumentation. In general I'm however inclined to think that a "second type of input file" meaning component assembly that is included in the file but not automatically rooted would be generally helpful. If we agree on implementing this option, I can easily modify SuperIlc to use it and drop this temporary hack. After all, I'm starting to get consistent R2RDump info for composite images so I believe I'm ready to dive into runtime implementation and after that the next step should be layered composite images I guess.

davidwrighton reviewed (changes requested):
I'd like us to have a discussion about how we expect arguments to be passed to the composite generator before we go down this route.

davidwrighton commented:
@trylek 
 About your points..
2) 10 minutes to compile 142 MB of code is ridiculous. I think we need to look at that, as that is the build mode we'll need for some of our hyperscale customers.
3) As noted in my review, please file an issue.
4) Could you please contact someone on the infrastructure or SDK side of the world? We're going to start having previews for .NET 5 in the not too distant future, and I'd like to make sure this is fixed when that happens.
5) Resources are a very interesting conundrum which we haven't explored in detail. Some resources could be merged, but most can't and it doesn't make sense. We may need to add some sort of scheme for recording fileversion data into the R2R format so that we can as developers understand what binaries are going into the merged binary.

trylek commented:
@davidwrighton - thanks a lot for your immediate feedback, here are my initial thoughts:

2) 10 minutes vs. 142 MB - I'll take a closer look what happens there. I've mostly been concentrating on functionality as opposed to performance right now so I haven't spun up any perf runs or instrumented experiments. I suspect that even the default PerfView analysis can provide more data as to what's the longest pole.

3) Filed https://github.com/dotnet/runtime/issues/31687;

4) Started an e-mail thread about this with Jeff and Tom;

5) Hopefully not blocking initially but something worth thinking about in more detail.

Thanks

Tomas

davidwrighton commented:
@trylek, I don't think you need to address the time to compile issue as part of this checkin, but that's a concern as we build out this feature. (Although if its a really obvious fix, feel free to stuff it in this changeset) I'm more interested to understand why you were merging all of the reference set into the input set in the general sense. That didn't make sense to me, and I'd like to understand your thinking there.

trylek commented:
@davidwrighton - for the 10 minutes compilation time, admittedly this is in debug mode I'm mostly using these days to debug the compiler; it's just 6 minutes when Crossgen2 itself is built in release mode and presumably even less with a release build of JIT so it's probably reasonable to leave more precise measurements for later.

For adding references to inputs, I guess that was more of an experiment rather than having any deep design foundations on my part - I've mostly been thinking that for typical apps we'd probably want something like this (only rooting the app and only compiling those parts of the framework found by the dependency analysis) rather than always carrying around the dead weight of the complete framework so I just tried that as that also lets me minimize the counterpart SuperIlc changes.

Ultimately, I tend to think that we should probably have two flavors of reference options in Crossgen2 to refer to "assemblies to be compiled into the composite file but without automatic rooting we're doing for input files" vs. "assemblies external to the composite file" (that will come into play once we start implementing layered composite files). At that moment we'll also need to figure out what to do about Crossgen2 options for actually compiling the framework - right now I'm not clear on whether in such case we should also root everything or for instance just the public API surface.

trylek committed:
Refactoring of Crossgen2 node factory / JIT interface for composite

This change seems to be the minimum letting me compile a small test
with the (unrooted) framework without crashes. It mostly deals
with proper propagation of the input module list to NodeFactory,
creating separate ModuleTokenResolver / SignatureContext for each
input module, adding the appropriate mapping logic and splitting
the R2R header nodes to global and assembly-level.

I'm not completely sure about the instance entrypoint table encoding,
that will need to come next as part of R2RDump correctness
validation. As next step I plan to make the required changes
to the PE header writer in the R2R executable emitter and
start working on R2RDump support. Another unfinished bit is
the handling of Win32 resources.

Thanks

Tomas

trylek committed:
Refactor R2R object writer for composite R2R executable support

1) Remove the remaining uses of PE reader from R2R PE writer;

2) Rename PEHeaderCopier to PEHeaderProvider as it can newly
either copy around an MSIL header or create a new one from scratch;

3) I have simplified the PEHeaderBuilder constructor by porting
over many of the hard-coded header constants from Crossgen1, avoiding
the need to copy them over from the input MSIL PE file.

4) I have cleaned up the logic around export symbols and leveraged it
for emitting the R2R_HEADER symbol.

5) Fixed a bunch of previously overlooked bugs around manifest
metadata table and the new assembly table.

Thanks

Tomas

trylek commented on src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:
```diff
@@ -2171,7 +2171,7 @@ private void getFieldInfo(ref CORINFO_RESOLVED_TOKEN pResolvedToken, CORINFO_MET
                             Debug.Assert(contextMethod.RequiresInstMethodDescArg());
                             helper = _compilation.NodeFactory.ReadyToRunHelperFromDictionaryLookup(
-                                helperId, runtimeDeterminedField.OwningType, contextMethod);
+                                helperId, runtimeDeterminedField.OwningType, contextMethod, _methodBeingCompiled.Context);
```
Thanks for spotting, reverted in 6th commit. I'll send out a separate PR for moving the getFieldInfo into the CorInfoImpl.ReadyToRun.cs source file, I agree that the conditional compilation structure has become quite complicated over time.

trylek replied:
This has already been reviewed and merged in separately.

trylek commented on src/coreclr/src/tools/Common/TypeSystem/Common/TypeSystemContext.cs:
```diff
@@ -45,13 +46,13 @@ public TargetDetails Target
         }
 
-        public ModuleDesc SystemModule
+        public EcmaModule SystemModule
```
Reverted in 6th commit, thanks for pointing that out.

trylek committed:
Initial Michal's PR feedback; SuperIlc support for composite build

I have addressed Michal's initial feedback - I reverted some
ECMA-specific changes from the common type system, I changed
HeaderNode to use a virtual method for the varying header part
and I removed the incorrect change to "non-READYTORUN" part of
getFieldInfo.

I have added provisions to SuperIlc to let me build tests in
composite mode; this mostly boils down to extending CompilerProcess
to support multiple input files and slightly restructuring the
build construction so that in composite mode we only run a single
compilation per folder.

As a small fix I noticed that the DLL name in the export table
is supposed to keep file extension.

Thanks

Tomas

trylek commented:
FYI, during my work on the counterpart R2RDump changes for the composite format I realized an interesting open question regarding import cells. I believe there are basically three ways to adjust import cells to fit the composite format and I'm interested in opinions on the preferable one:

1) The current working version randomly collects all import cells. This means they are heterogeneous - their signatures are encoded relative to different caller modules. Import cells are potentially duplicated even though actual local methods aren't - two different submodules of the composite R2R generally need two different encodings of a given target.

2) Keep module-relative import cells but separate them into per-module import sections. This is easier for R2RDump and perhaps even for data access locality; there's still duplication in encoding the same import cell relative to two different modules.

3) Change the runtime signature behavior for composite R2R to avoid import cell duplication, either by enforcing the presence of module override on each signature or by arbitrarily choosing a "root" module and encoding / decoding all signatures within the composite file relative to this root module. 

Choice of the root module doesn't matter - manifest metadata covers the entire version bubble of the composite image so that we can encode signatures relative to any of the modules present, a smart choice might just yield a minor size optimization in making some of the signatures a byte or two shorter.

jkotas commented:
3 - "Change the runtime signature behavior for composite R2R to avoid import cell duplication" looks like the best choice.

trylek commented:
Thanks Jan, I agree with you, I have started implementing option 3. In the meantime I have found another funny thing: in my current implementation I'm using the manifest metadata to represent the assemblies within the composite R2R file. After implementing export table directory support in R2RDump and becoming able to parse RTR_HEADER, I've immediately hit a new issue: as we're using AssemblyRef entries to represent the reference assemblies, this naturally doesn't work when the assembly name differs from the module name. I hit this just by lucky chance as I'm still using the regression test b426654 for the initial experiments and that's exactly an IL-written test where the module is named "b426654" but the assembly name is "cs1". I'm wondering whether we should use ModuleRefs instead in the composite case.

jkotas commented:
We do hot have multimodule assemblies in .NET Core. Why does this matter?

trylek commented:
Well, it matters at least for CoreCLR tests that involve IL assemblies with hand-written assembly vs. module names. We may decide to not care about them but I'm worried it will make our life harder w.r.t. testing and it seems to me like an overal debt in the implementation.

jkotas commented:
Ok, but how does it matter for them? I am not following.

jkotas commented:
The assembly name should be used for both assembly ref and assembly def. The filename should not need to be stored anywhere.

trylek commented:
OK, you're probably right it's merely an annoyance for R2RDump, not for the runtime as such, it just sucks that we cannot use the manifest metadata to easily identify the appropriate standalone MSIL files corresponding to the composite R2R components. For the "b426654" test the main app module entry in the manifest metadata is "cs1" so I'll need to try to open all managed modules in all reference paths looking for the one "r426654" that happens to contain the assembly "cs1".

jkotas commented:
> I'll need to try to open all managed modules in all reference paths looking for the one "r426654" that happens to contain the assembly "cs1".

I do not think you need to do that. You can just declare this case unsupported. There are number of places in the .net tooling that assume assembly name == file name.

trylek commented:
OK, that sounds reasonable or at least reasonably simplifying for now. Thanks Jan for sharing your detailed feedback!

trylek committed:
Initial R2RDump support for composite images

With this change I have managed to get the first seemingly
meaningful dump of a composite R2R executable. One remaining
TODO I haven't yet managed to fully fix is EH clauses that
contain class tokens for typed catch clauses. I'm not yet
completely clear on how easy it is to get back from the EH
clause to the containing method / module so Im thinking that
perhaps in composite case we could change the interpretation
of the class token dword to instead point into a special table
mapping these dwords to the ordered pairs (module, token).

I have also changed the code factory so that input module
ordering is explicit which is needed in multiple places.
I have fixed manifest metadata so that entries for all
components assemblies are generated upfront in the proper
order to match the component assembly table entry indices.
In signature and instance entrypoint emitter I have made sure
to always generate module prefixes in composite mode in order
to make their interpretation unique across all component
assemblies. I have yet to slightly mop up the context and
resolver processing as it currently still causes duplication
of import cells emitted from the different modules.

In R2RDump I have written a new piece of code to parse
the export table directory to get to the RTR_HEADER
ReadyToRun header pointer and use it to parse the component
assembly table and per-assembly type and method entrypoint tables.

Thanks

Tomas

trylek committed:
Simplify resolver & signature context management

Based on my exchange with JanK on the PR thread I realized I can
substantially simplify signature context and module token resolver
management in Crossgen2: as we don't want to produce
module-specific import cells, we also don't need multiple
signature context and multiple resolvers.

I have basically reverted a part of my change implementing the new
class InputModule - without the need for module-specific context
and resolvers we can basically return to the old way of having
just one resolver and context in the node factory.

The new logic is that, in the composite R2R case, we create a special
"null signature context" that subsequently forces the emission
of explicit module overrides on all signatures produces on the
import cells and thus makes them module-agnostic.

Thanks

Tomas

trylek committed:
Previously missed newly added R2RDump source file PEReaderExtensions

trylek commented:
OK, with the latest changes I seem to be getting consistent R2RDump info for composite images (except for the EHClauses that are currently commented out due to a design bit I have yet to clarify). I believe I'm ready to start working on the relevant runtime changes but I'd prefer to review and merge in the existing changes separately as the change is already big. I'll start out by sending out a separate review for the R2RDump parts of this change that are relatively straightforward. I'll follow up by preparing separate PR's for SuperIlc and Crossgen2 changes after applying @davidwrighton's feedback towards my "non-rooting input" suggestion.

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs:
```diff
@@ -57,6 +58,12 @@ public sealed class NodeFactory
 
         public NameMangler NameMangler { get; }
 
+        public bool Composite { get; }
```
`CompilationModuleGroup` might be a good place to hang these three since it already knows which modules are in compilation. I know we've tried to keep non-node state out of `NodeFactory` in the past to stop it becoming a dumping ground.

trylek replied:
Good point, fixed.

trylek committed:
Simon's PR feedback and general cleanups

I have moved the 'Composite' flag into CompilationModuleGroup
(under the name IsCompositeBuildMode). For CompilationModuleSet,
I just changed it to an enumerable of EcmaModule, with my most
recent resolver / context changes it doesn't really need to have
such a complex structure.

Based on David Wrighton's feedback I rolled back my experimental
changes altering the behavior of reference assemblies in composite
mode and instead I introduced a new Crossgen2 option '-u' to
supply "unrooted input files"; I have modified SuperIlc to use this
option when invoking Crossgen2 in composite build mode.

Thanks

Tomas

trylek committed:
Rewriting standalone MSIL assemblies with forwarder info injection

This change modifies Crossgen2 so that, during composite R2R
compilation, it rewrites the individual input MSIL assemblies to
the output folder, marking them with a special type of R2R header
stating that this is a component assembly compiled into a composite
R2R image (by means of a new section OwnerCompositeExecutable).

I have also renamed Context in NodeFactory to SignatureContext to
avoid ambiguity with TypeSystemContext. In the R2R compilation
the rewriting logic merits additional cleanup - creating of the
node graph for component assemblies should be probably moved into
the node factory and consolidated with AttachToDependencyGraph
to reduce duplication. (But the actual duplication isn't huge.)

I have also added provisions to R2RDump to honor this new section
type so that it can dump the contents of the rewritten standalone
MSIL.

Thanks

Tomas

trylek committed:
Simplify signature context management

As based on my recent design discussions with JanK we want the
import cells in composite R2R images to be unique across all modules,
we don't need different signature contexts for emitting them. This
means that we no longer need to back-propagate the signature context
from CorInfoImpl to R2R codegen node factory.

Thanks

Tomas

trylek committed:
Rebase against my signature cleanup change

trylek commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/IBC/IBCProfileData.cs:
```diff
@@ -15,10 +15,10 @@ public IBCProfileData(bool partialNGen, IEnumerable<MethodProfileData> methodDat
-
-                _methodData.Add(data.Method, data);
+                if (!_methodData.ContainsKey(data.Method))
+                {
```
Filed https://github.com/dotnet/runtime/issues/31687.

trylek commented:
OK, I think it doesn't make sense for me delaying this review anymore especially as I recently found out that DavidWr and JanV are both on vacation next week. I have verified that the changes don't break "non-composite" Crossgen2 build and that they are consistent enough so that R2RDump is capable of dumping some composite images. I'm currently working on the runtime support but that's going to be easier to review separately anyway.

trylek committed:
Fix PE header characteristics to make the images loadable by runtime

trylek commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/HeaderNode.cs:
```diff
@@ -106,12 +107,15 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
 
-            // ReadyToRunHeader.Magic
-            builder.EmitInt((int)(ReadyToRunHeaderConstants.Signature));
+            if (_isGlobalHeader)
```
Good point, fixed.

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:
```diff
@@ -79,10 +81,39 @@ public void EmitPortableExecutable()
+                        Subsystem.Unknown,
+                        _nodeFactory.Target);
+                    timeDateStamp = 0;
+                    r2rHeaderExportSymbol = _nodeFactory.Header;
```
We should never force the PE timestamp to zero. That will cause problems with getting a pdb to work. Instead, we should either copy a timestamp, as we do already for the non-composite scenario, or use a deterministic id provider. (Basically, I'm arguing that this timestamp parameter should be passed as a nullable int, and this case should be null.

davidwrighton replied:
And not write a new timestamp. 

Or, we should implement a deterministic id provider, and capture the timestamp out of that. 

The general path for that is:
Func<IEnumerable<Blob>, BlobContentId> deterministicProvider = new new Func<IEnumerable<Blob>, BlobContentId>(content => BlobContentId.FromHash(CryptographicHashProvider.ComputeHash(HashAlgorithmName.SHA1, content));

then pass that provider to the PEBuilder constructor. See PeWriter.cs in the Roslyn codebase for details, or ask the Roslyn folks for advise on how to do that.

---
In reply to: [379591234](https://github.com/dotnet/runtime/pull/31663#discussion_r379591234) [](ancestors = 379591234)

trylek replied:
I have created the following tracking issue for this: https://github.com/dotnet/runtime/issues/32507

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/ObjectWriter/R2RPEBuilder.cs:
```diff
@@ -270,16 +260,17 @@ public int GetSymbolFilePosition(ISymbolNode symbol)
         /// </summary>
         /// <param name="outputStream">Output stream for the final R2R PE file</param>
-        public void Write(Stream outputStream)
+        public void Write(Stream outputStream, int timeDateStamp)
```
>int [](start = 47, length = 3)

This should be nullable or something to indicate that we don't want to force the timestamp to be reset here.

trylek replied:
Well, I think it's more about getting the proper <code>timeDateStamp</code> in the first place, we need to write something into the PE header.

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/ObjectWriter/R2RPEBuilder.cs:
```diff
@@ -401,9 +392,10 @@ private void ApplyMachineOSOverride(Stream outputStream)
         /// <param name="outputStream">Output stream representing the R2R PE executable</param>
-        private void CopyTimeStampFromInputImage(Stream outputStream)
+        /// <param name="timeDateStamp">Timestamp to set in the R2R PE header</param>
+        private void CopyTimeStampFromInputImage(Stream outputStream, int timeDateStamp)
```
>CopyTimeStampFromInputImage [](start = 21, length = 27)

Rename this method, and update the comment. Its no longer copying a timestamp, its simply setting it.

trylek replied:
Good point, fixed.

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/ObjectWriter/R2RPEBuilder.cs:
```diff
@@ -560,25 +557,32 @@ static class PEHeaderCopier
-                imageCharacteristics &= ~Characteristics.Bit32Machine;
-                imageCharacteristics |= Characteristics.LargeAddressAware;
-            }
+            characteristics |= (is64BitTarget ? Characteristics.LargeAddressAware : Characteristics.Bit32Machine);
```
This isn't doing the same as the code it replaces. In particular, the old code when presented with a Bit32Machine entry was clearing it.

trylek replied:
I believe the proper way to set these bits is to have them mutually exclusive - i.e. in 32-bit mode, LargeAddressAware is cleared and Bit32Machine is set and vice versa in 64-bit mode. IIRC LargeAddressAware + Bit32Machine is the combo used for AnyCPU MSIL but that doesn't make sense for the output R2R native image. I have modified the change in this sense, thanks for pointing out this inconsistency!

davidwrighton commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/ObjectWriter/R2RPEBuilder.cs:
```diff
@@ -596,11 +600,13 @@ public static PEHeaderBuilder Copy(PEHeaders peHeaders, TargetDetails target)
 
-            // Copy over selected DLL characteristics bits from IL image
-            dllCharacteristics |= peHeaders.PEHeader.DllCharacteristics &
-                (DllCharacteristics.TerminalServerAware | DllCharacteristics.AppContainer);
```
In general, the changes in here look meaningful, and I'm not sure why you made them. Please work with Simon to make sure these are the right changes to make.

nattress replied:
I use https://github.com/dotnet/runtime/blob/b8591b4a9ce88f98ff6aba93241f23e1512c01fa/src/coreclr/src/zap/zapimage.cpp#L1511 as the source of truth for these values so we behave the same as Crossgen1

trylek replied:
I have compared the Crossgen2 code to the bit in zapimage.cpp you pointed out and I fixed a few small differences but I believe that even before we were mostly doing the same thing.

davidwrighton commented:
@trylek, I've approved the changes, assuming you address my comments. I'm going on vacation, so I didn't want to leave a disapproval here, but please get someone else to approve before checkin too.

trylek commented:
@davidwrighton - thanks for your detailed feedback, I'll make sure to work with Simon on ironing out the remaining wrinkles and making sure we emit reasonable PE header information. Enjoy your vacation!

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/CodeGen/ReadyToRunObjectWriter.cs:
```diff
@@ -26,8 +28,8 @@ internal class ReadyToRunObjectWriter
         // Nodefactory for which ObjectWriter is instantiated for.
         private readonly NodeFactory _nodeFactory;
         private readonly string _objectFilePath;
+        private readonly EcmaModule _componentModule;
```
Could you add a comment as to what `_componentModule` is for and when we should expect it to be null or not. It'll make it easier to reason about this member in the future.

trylek replied:
Good point, added XML comments for all the fields.

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/InstanceEntryPointTableNode.cs:
```diff
@@ -52,9 +56,11 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
-                    bool enforceOwningType = false;
-                    ModuleToken moduleToken = factory.SignatureContext.GetModuleTokenForMethod(method.Method.GetTypicalMethodDefinition());
-                    if (moduleToken.Module != factory.SignatureContext.GlobalContext)
+                    // In composite R2R format, always enforce owning type to let us share generic instantiations among modules
```
Nit: This check feels like a policy decision we should ask from the ModuleGroup so decisions based on composite-ness are centralized

trylek replied:
Good point. I have added a new method <code>EnforceOwningType(EcmaModule module)</code> to CompilationModuleGroup.

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/ManifestMetadataTableNode.cs:
```diff
@@ -62,37 +62,53 @@ public class ManifestMetadataTableNode : HeaderTableNode
-                _assemblyRefToModuleIdMap[assemblyName] = assemblyRefIndex;
+                if (_nodeFactory.CompilationModuleGroup.CompilationModuleSet.Skip(1).Any())
+                {
+                    throw new InternalCompilerErrorException("Multiple input files specified without the composite flag");
```
Nit: This feels like something we should catch when setting up the `CompilationModuleGroup` so we can emit a localized error to the console and maybe just assert here.

trylek replied:
Good point, fixed.

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/OwnerCompositeExecutableNode.cs:
```diff
@@ -0,0 +1,39 @@
+
+namespace ILCompiler.DependencyAnalysis.ReadyToRun
+{
+    internal class OwnerCompositeExecutableNode : HeaderTableNode
```
Would you mind adding a comment describing this? In multi-assembly-out composite mode, this is used by the non-master IL assemblies to point at the main composite R2R assembly?

trylek replied:
Sure, fixed.

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs:
```diff
@@ -174,28 +176,39 @@ public override int GetHashCode()
+            {
+                if (CompilationModuleGroup.CompilationModuleSet.Skip(1).Any())
+                {
+                    throw new InternalCompilerErrorException($"Multiple input modules ({CompilationModuleGroup.CompilationModuleSet.Count()}) are only supported in composite R2R build mode");
```
Nit: Make this an assert and have the real check when we're setting up compilation and can write an error out to the user?

trylek replied:
I have changed the LINQ expression to use <code>Single()</code> to throw when the number of elements is not 1.

nattress commented on src/coreclr/src/tools/crossgen2/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:
```diff
@@ -206,40 +208,87 @@ public sealed class ReadyToRunCodegenCompilation : Compilation
 
+        private void RewriteComponentFile(string inputFile, string outputFile, string ownerExecutableName)
+        {
+            EcmaModule inputModule = NodeFactory.TypeSystemContext.GetModuleFromPath(inputFile);
```
I like this method! - it shows how flexible our system is to add a thing to a module and save it in ~20 lines of code.

trylek replied:
Me too :-).

nattress reviewed (approved):
LGTM - I made a couple nits about comments

trylek committed:
Address easy parts of Simon's PR feedback (no PE header changes yet)

trylek commented:
Thanks Simon for your feedback. In the latest (15th) commit I have addressed "conservative" parts of your PR feedback like improved comments and minor code reorganizations. I haven't yet addressed PE header feedback including DavidWr's feedback. I'm not sure how to approach the deterministic ID thing, it looks like a new functionality we don't have any support for in the compiler just yet.

My current plan is to just play with the runtime over the weekend and hopefully get to the point I'm able to actually call [non-generic] methods in the composite R2R executable. Ideally early next week I should be able to review and merge in the initial support so that we can fan out further work like shared framework support, diagnostic support, additional performance tuning knobs etc.

jkotas commented:
> it looks like a new functionality we don't have any support for in the compiler just yet.

So open a new issue on it and address it in separate PR.

trylek committed:
Fixes for composite build mode based on initial runtime debugging

By debugging the CoreCLR runtime running the composite R2R image
I found found that available types and inlining info wasn't getting
filtered by current module. Due to this fact both tables were
basically getting duplicated 240 times for each of the component
modules. Fixing this reduced the size of the composite R2R image
for a small JIT test from 75 MB to 1.5 MB.

Thanks

Tomas

trylek committed:
Address remaining PR feedback from David and Simon

trylek committed:
Fix attribute presence table to be per assembly
